src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -1508,5 +1508,279 @@ public void nameof(string x)
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotReferenceInstanceMembersFromConstructorInitializersInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C(string s){}
+    public C() : this(nameof(S.Length)){}
+    public string S { get; }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (5,30): error CS8652: The feature 'reduced member access checks in 'nameof'' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public C() : this(nameof(S.Length)){}
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "S").WithArguments("reduced member access checks in 'nameof'").WithLocation(5, 30));
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotAccessStructInstancePropertyInLambdaInNameofInCSharp9()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public string P { get; }
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(P.Length);
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (9,42): error CS8652: The feature 'reduced member access checks in 'nameof'' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         Func<string> func = () => nameof(P.Length);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "P").WithArguments("reduced member access checks in 'nameof'").WithLocation(9, 42));
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotReferenceStaticPropertyFromInstanceMemberInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static int StaticProp { get; }
+    public string M() => nameof(Prop.StaticProp);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (6,33): error CS8652: The feature 'reduced member access checks in 'nameof'' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public string M() => nameof(Prop.StaticProp);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "Prop.StaticProp").WithArguments("reduced member access checks in 'nameof'").WithLocation(6, 33));
+        }
+
+        [Fact]
+        public void TestCanReferenceStaticMethodFromInstanceMemberInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static void StaticMethod(){}
+    public string M() => nameof(Prop.StaticMethod);
```

Comment:
Consider adding some of these used recursively. For example:
```cs
class C
{
    [Attr(nameof(Prop.StaticMethod))]
    public C Prop { get; }
    public static void StaticMethod(){}
}
class Attr : Attribute { public Attr(string s) {} }
```
Other similar instances of attribute recursion would be good to test as well.

Reply:
Also, please verify Emit diagnostics for these, to make sure we catch any potential cycles all the way down.

---
In reply to: [549462200](https://github.com/dotnet/roslyn/pull/48754#discussion_r549462200) [](ancestors = 549462200)